--- boost/format/alt_sstream_impl.hpp
+++ boost/format/alt_sstream_impl.hpp
@@ -258,7 +258,7 @@
                     void *vdptr = alloc_.allocate(new_size, is_allocated_? oldptr : 0);
                     newptr = static_cast<Ch *>(vdptr);
 #else
-                    newptr = alloc_.allocate(new_size, is_allocated_? oldptr : 0);
+                    newptr = std::allocator_traits<compat_allocator_type>::allocate(alloc_, new_size, is_allocated_? oldptr : 0);
 #endif
                 }
 
