From 1c762309166c3e5489575a9337cdd517eea97e45 Mon Sep 17 00:00:00 2001
From: Tobias Markus <tobbi.bugs@googlemail.com>
Date: Wed, 24 Dec 2025 01:47:55 +0100
Subject: [PATCH] Add pkgconf as a normal dependency

---
 ports/libraqm/vcpkg.json | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ports/libraqm/vcpkg.json b/ports/libraqm/vcpkg.json
index 5ac3d2a9f55b31..7debcc00ce21e1 100644
--- a/ports/libraqm/vcpkg.json
+++ b/ports/libraqm/vcpkg.json
@@ -11,6 +11,7 @@
     {
       "name": "vcpkg-tool-meson",
       "host": true
-    }
+    },
+    "pkgconf"
   ]
 }
