diff --git a/external/squirrel/squirrel/sqlexer.cpp b/external/squirrel/squirrel/sqlexer.cpp
index 547835b..5d47dfe 100644
--- a/external/squirrel/squirrel/sqlexer.cpp
+++ b/external/squirrel/squirrel/sqlexer.cpp
@@ -148,11 +148,11 @@ SQInteger SQLexer::Lex()
 			case _SC('='):
 				NEXT();
 				RETURN_TOKEN(TK_DIVEQ);
-				continue;
+				break;
 			case _SC('>'):
 				NEXT();
 				RETURN_TOKEN(TK_ATTR_CLOSE);
-				continue;
+				break;
 			default:
 				RETURN_TOKEN('/');
 			}
